✨ Combobox

larseirikhansen opened the pull request.

larseirikhansen committed:
fix Meta import error storybook

larseirikhansen committed:
only useEffect triggers focusInput

larseirikhansen committed:
made list absolute to input wrapper

larseirikhansen committed:
fixed immediate focus bug by adding usePrevious hook

larseirikhansen committed:
remove comment

larseirikhansen committed:
WIP: trying to fix focus, see TODOS in Combobox.tsx for reference

it-vegard committed:
Prevent moving cursor in combobox input when moving between options in dropdown list

larseirikhansen committed:
focus is mvp

it-vegard committed:
Expose to screen readers which options in combobox is active or selected

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
New design for checked options in combobox

it-vegard committed:
Add separate focus styling for combobox dropdown, to separate it from selected options

it-vegard committed:
Simplify focus management for combobox

Handle open/close dropdown on focus/blur for the entire component, not at each interactive element inside it.
This works because events bubble up, causing onFocus/onBlur on outer div to be called.

it-vegard committed:
Close combobox dropdown on escape, and re-open it when typing

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
Fix re-rendring of all chips when one is added or removed. This affected the combobox component so it temporarily lost focus from the input field.

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
Add option for "add new value" in combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
refactor events, fix empty chip bug

larseirikhansen committed:
fix click-to-add, working on add-new option

larseirikhansen committed:
rm logs, fix dependency array

larseirikhansen committed:
wip before changing state logic to reducer

larseirikhansen committed:
small refactoring

larseirikhansen committed:
remove reducer

larseirikhansen committed:
splitting options and selected options into different files, working afaik

larseirikhansen committed:
refactor clear button and toggle list button into new files

larseirikhansen committed:
Merge 'origin/main' into combobox

larseirikhansen committed:
scroll to child in filteredlist when using arrows

larseirikhansen committed:
remove TODO comments

it-vegard committed:
Renamed internalValue -> value in FilteredOptions

larseirikhansen committed:
fix event race condition bug

larseirikhansen committed:
small reorganizing of code

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

it-vegard committed:
Use proper rounded borders around combobox, and actually show border around the dropdown list

larseirikhansen committed:
WIP: create a 'just-added' list, make memoized variable combining internal and external values

larseirikhansen committed:
add just-added list, a semi-persistant list with custom values

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
add more options, add preselected options

larseirikhansen committed:
update storybook to rc.5

larseirikhansen committed:
more todos

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
add hook for custom options

larseirikhansen committed:
remove native clear btn, useFilteredOptions hook

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
...

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
refactor filteredoption state, context-based file structure

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
moved icon

larseirikhansen committed:
custom options context

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
contexts for selectedOptions, filteredOptions, customOptions

larseirikhansen committed:
var name change

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
Merge branch 'main' into combobox

it-vegard committed:
Refactor wrapper component of Combobox into a separate component

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

it-vegard committed:
Split code for input value and onchange in combobox into a separate hook

it-vegard committed:
Created InputContext

it-vegard committed:
Correctly set "add new"-option as activedescendant for combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
yarn lock updates

larseirikhansen committed:
mobile version

it-vegard committed:
Move virtual focus-logic for FilteredOptions from Combobox to FilteredOptionsContext, so it is easier to act on internal FilteredOptions-state

it-vegard committed:
Handle virtual focus for combobox switching between input field, add new option-button and the FilteredOptions list.

Needs some more styling for add new option.

it-vegard committed:
Fix TS-errors

it-vegard committed:
Remove aria-activedecendant when no FilteredOptions item has virtual focus

it-vegard committed:
Use arrow up/down to move virtual focus into and inside of FilteredOptions list only if the keystrokes should not move the cursor in the text field

it-vegard committed:
Logic for checking if the value should be considered a new option and logic for filtering options based on the value had been mistakenly combined.

This fixes issues with "add new"-button and filtering

it-vegard committed:
Fix issue with selecting/deselecting the first option in FilteredOptions

it-vegard committed:
Prevent flickering focus for "add new" button in FilteredOptions

larseirikhansen committed:
remove comments

larseirikhansen committed:
merge commit

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
yarn lock changes

larseirikhansen committed:
remove unused props, add loading prop

larseirikhansen committed:
rm comments

larseirikhansen committed:
add focus-shadow to open/close-button, styles to loading option

larseirikhansen committed:
set up story with perpetually loading data

larseirikhansen committed:
styling loading list item

larseirikhansen committed:
input in different file to open for <input> outside of selectedOptions

it-vegard committed:
Added a working test for combobox

it-vegard committed:
We never use setOptions internally, so just removed all references to it. Options should just be a prop, which can change from the outside.

it-vegard committed:
Fix aria-controls, as it should always reference the list of options

it-vegard committed:
Add test for combobox loading state

it-vegard committed:
Make selectedOptions prop optional and replace setSelectedOptions with an onToggleSelected callback

We believe a callback function is more useable than having to replicate the internal state handling with the setSelecetedOptions function.
Making selectedOptions optional, we enable external chips, for multi-filter pages, etc.

it-vegard committed:
Improve handling of virtual focus and selection

it-vegard committed:
Added example of combobox with callback functions

it-vegard committed:
Merge branch 'main' into combobox

# Conflicts:
#	yarn.lock

larseirikhansen committed:
umcomment example

it-vegard committed:
Un-comment file

larseirikhansen committed:
WIP: singleSelect

larseirikhansen committed:
...

it-vegard committed:
Fix non-new item selection when add new is displayed

it-vegard committed:
Update yarn.lock

it-vegard committed:
Replaced references to ds-icons with new aksel-icons

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

it-vegard committed:
Trigger new build

it-vegard committed:
Testing: Removing all aria, so we can add incrementally to see when braille breaks

it-vegard committed:
Reverting previous change, but removes aria-owns and changes type="search" to type="text"

it-vegard committed:
Remove aria-invalid and variant props from combobox input attributes

it-vegard committed:
Remove the toggle button from tab order, as it does the same as arrow navigation and typing in the input field. Also added state (expanded) and modified the accessible name

it-vegard committed:
Add button-style focus to "Add new" button in Combobox FilteredOptions list

it-vegard committed:
Remove focus indicator from component when navigating FilteredOptions list. Just an idea, that needs to be checked/accepted by design

larseirikhansen committed:
wip single select

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
wip mvp singleSelect

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
wip: singleSelect create searchTerm

it-vegard committed:
Re-add missing id for input field to fix tests and accessibility issue with connecting label and input

it-vegard committed:
New yarn.lock with updated aksel dependencies

it-vegard committed:
Add missing labels and id attributes in Storybook

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
yarn lock update

it-vegard committed:
Merge branch 'main' into combobox

# Conflicts:
#	yarn.lock

KenAJoh commented on @navikt/core/css/accordion.css:
Ser ut som det er kjørt custom prettier config på en del filer. Beste er om samme config som ligger i `package.json`  brukes på all koden.

it-vegard replied:
Fjernet printWidth i package.json, som gjorde at vi fikk bort diffen i urørte filer

KenAJoh commented on @navikt/core/react/src/form/combobox/SelectedOptions/SelectedOptions.tsx:
```diff
@@ -0,0 +1,38 @@
+            <Chips.Removable
+              className="navds-combobox__selected-option"
+              key={option + i}
+              onMouseUp={(e) => {
```
Tror denne kan byttes ut med `onPointerUp` for enda bedre device-støtte (https://javascript.info/pointer-events)

it-vegard replied:
Det kunne den. Har gjort det alle steder onMouseXXX ble brukt (scopet til combobox-mappen).

it-vegard committed:
Fix broken screen reader/braille display for combobox after refactoring Input field

Fix incorrect IDs/ID pointers (aria-activedescendant++)

it-vegard committed:
Add return type for useFormField hook

it-vegard committed:
Fix WithExternalChips example - missed selectedOptions prop

larseirikhansen committed:
merge from main

it-vegard committed:
Working code for single select with autocomplete

it-vegard committed:
Fix ts error for onchange event

it-vegard committed:
Clean up props/types for combobox

it-vegard committed:
Forgot to remove old, unused props

it-vegard committed:
Missing label

it-vegard committed:
Feeding the single select prop to the component through the selectedOptionsContext

it-vegard committed:
Merge branch 'main' into combobox

it-vegard committed:
Split single select into one version with autocomplete and one without (for testing purposes)

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
Fix incorrect id in aria-controls attribute

it-vegard committed:
Add aria-describedby for autocomplete to get screen readers to read out the full word that is autocompleted

it-vegard committed:
Fix nullpointer exception

it-vegard committed:
Added hotkeys Home/End to Combobox, moving focus to input field or last element in dropdown list, respectively.

it-vegard committed:
Rename prop loading -> isLoading

it-vegard committed:
Announce isLoading state for Combobox by connecting it to the input field using aria-describedby

it-vegard committed:
Add prop for allowing new values. Default off.

it-vegard committed:
Clean up onChange signature (use native event as parameter)

it-vegard committed:
Add feedback in FilteredOptions for no hits.

it-vegard committed:
Replace onMouseXXX with onPointerUp for consistency and better device support

it-vegard committed:
Fix broken test after renaming prop

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

KenAJoh committed:
Merge branch 'main' into combobox

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+            "navds-combobox__wrapper-inner navds-text-field__input",
+            {
+              "navds-combobox__wrapper-inner--virtually-unfocused":
+                currentOption,
```
Som edgecase kan `currentOption` være `""` -> `Boolean("") === false` 
```suggestion
                currentOption !== null,
```

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+                {...rest}
+              />
+            </>
+          ) : (
```
```suggestion
        >
          {singleSelect ? (
            <Input
              id={inputProps.id}
              key="combobox-input"
              ref={mergedInputRef}
              inputClassName={inputClassName}
              handleClear={handleClear}
              toggleOption={toggleOption}
              singleSelect
              {...rest}
            />
          ) : (
```

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -0,0 +1,110 @@
+            aria-selected={false}
+            id={`${id}-is-loading`}
+          >
+            <Loader />
```
Bør Loader ha `aria-hidden` her?

JulianNymark replied:
Jeg er usikker på hvordan screen-reader experience er best for veldig kortvarige ting? (jeg ville gjettet at ingenting er bedre en kortvarig støy? dvs ja til `aria-hidden` 🤔 ). Ville kanskje spurt Morten hva han tenker? Godt spørsmål!

it-vegard replied:
Tror vi har testet det med Morten, men er ikke helt sikker.

I utgangspunktet er jeg litt skeptisk til å skjule den: isLoading er en prop satt på komponenten for tilfeller der det tar tid å få svar på søket. Foreløpig har vi derfor eksplisitt koblet den til input med aria-describedby.

Men kan godt ta en runde på det på mandag. Hvis vi skal skjule den, så gjør vi det bedre med å ikke bruke aria-describedby i dette tilfellet.

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -0,0 +1,110 @@
+              !selectedOptions.includes(value)
+            } /* TODO: Should this attribute ever be true? Can the add-button have the selected state? */
+          >
+            <PlusIcon />
```
```suggestion
            <PlusIcon aria-hidden />
```

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -0,0 +1,110 @@
+            } /* TODO: Should this attribute ever be true? Can the add-button have the selected state? */
+          >
+            <PlusIcon />
+            <BodyShort size="medium">
```
```suggestion
            <BodyShort>
```

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -0,0 +1,110 @@
+            tabIndex={-1}
+            onClick={(e) => {
+              toggleOption(e);
+            }}
```
```suggestion
            onClick={toggleOption}
```

it-vegard replied:
Har lagt til option som et argument her, så denne er ikke lenger relevant

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/FilteredOptions.tsx:
```diff
@@ -0,0 +1,110 @@
+            role="option"
+            aria-selected={selectedOptions.includes(o)}
+          >
+            <BodyShort size="medium">{o}</BodyShort>
```
```suggestion
            <BodyShort>{o}</BodyShort>
```

KenAJoh commented on @navikt/core/react/src/form/combobox/FilteredOptions/filteredOptionsContext.tsx:
```diff
@@ -0,0 +1,278 @@
+  useEffect(() => {
+    if (isExternalListOpen !== undefined)
+      setInternalListOpen(isExternalListOpen);
+  }, [isExternalListOpen]);
```
```suggestion
  useEffect(() => isExternalListOpen !== undefined && setInternalListOpen(isExternalListOpen)

  , [isExternalListOpen]);
```

it-vegard replied:
Replaced this useEffect with a useMemo, that evaluates to `isExternalListOpen ?? isInternalListOpen`.

KenAJoh commented on package.json:
```diff
@@ -34,7 +34,7 @@
       {
         "files": "**/*.css",
         "options": {
-          "printWidth": 130
```
Er sikkert denne som har påvirket accordion, button, expansioncard og flere CSS-filer. Kan kanske vært en egen PR hvis dette er noe vi ønsker å endre på. Må også da kjøre pretter på alle filene i systemet for å synke de med denne verdien

it-vegard replied:
Stemmer. Har fjernet den og dermed diffene i filene som ble påvirket.

KenAJoh commented on @navikt/core/react/src/form/combobox/ClearButton.tsx:
```diff
@@ -0,0 +1,27 @@
+  );
+};
+
+export default ClearButton;
```
Out-of-scope for denne PR-en sikkert, men kan "ClearButton" brukt i `<Search />` også trekkes ut slik at begge bruker samme komponent. Tok bare en raskt titt, men tror API-et er relativt likt mellom dem (ikke sikker på designet)

it-vegard replied:
Enig, men i en egen PR seinere? 😅

KenAJoh replied:
Ja kan være en TODO for fremtiden haha 👍

KenAJoh committed:
:arrow_up: yarn lock

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+
+  // we have the code for error handling, and dont do anythign with it
+  // TODO: if text is long, new line
+  // TODO: mobile, should fewer options be shown at a time?
```
```suggestion
```

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+  );
+});
+
+// TODO: Remove this line. Only added to trigger build
```
```suggestion
```

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+import ComboboxWrapper from "./ComboboxWrapper";
+import { useInputContext } from "./Input/inputContext";
+import Input from "./Input/Input";
+
```
Kan legge til JSdoc her (se andre komponenter)

it-vegard replied:
Kan være vi bør se litt på strukturen vår her (det vi eksponerer som Combobox er egentlig ComboboxProvider).
Men har lagt til JSDoc i types.ts, men skal få lagt inn noe for selve komponenten også.

it-vegard replied:
Har skrevet en liten JSDoc for komponenten (i ComboboxProvider, som er den vi eksporterer) med kodeeksempel.

KenAJoh commented on @navikt/core/react/src/form/combobox/Combobox.tsx:
```diff
@@ -0,0 +1,247 @@
+
+  //focus on input whenever selectedOptions changes
+  useEffect(() => {
+    if (prevSelectedOptions !== selectedOptions) focusInput();
```
Denne funksjonaliteten kan flyttes inn i `useSelectedOptionsContext` ([ref](https://react.dev/learn/you-might-not-need-an-effect)). Focus vil settes etter en ekstra render-syklus nå, som i worst-case kan virke som systemet er tregt
```
const {
    selectedOptions,
    prevSelectedOptions,
    removeSelectedOption,
    addSelectedOption,
    singleSelect,
  } = useSelectedOptionsContext(inputRef);

```

it-vegard replied:
Done

KenAJoh commented on @navikt/core/react/src/form/combobox/SelectedOptions/selectedOptionsContext.tsx:
```diff
@@ -0,0 +1,103 @@
+  useEffect(
+    () => setSelectedOptions(externalSelectedOptions || []),
+    [externalSelectedOptions]
+  );
```
```suggestion
if(selectedOptions !== externalSelectedOptions){
   setSelectedOptions(externalSelectedOptions || [])
}
```
https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes
Ifølge React vil dette være den best måten å lytte til eksterne props. Virker litt skummelt å ha rene if-tester med setState rett i koden, men er det react ønsker 🤷

it-vegard replied:
Eksempelet til React har et ekstra element som ser viktig ut her, nemlig prevItem/setPrevItem.
Uten den får vi ikke gjort en sjekk på om externalSelectedOptions har endret seg.

Men, de lister jo opp det de mener er en enda bedre løsning, nemlig å bare "kalkulere verdien under rendering".
Det ser ut til å fungere hos oss, med følgende kode:

`const [internalSelectedOptions, setSelectedOptions] = useState<string[]>([]);
 const selectedOptions = externalSelectedOptions ?? internalSelectedOptions;`
 
 Vurderte en useMemo i stedet, men her er det jo bare snakk om å velge hvilken verdi man skal bruke. Enten bruker vi den eksterne, eller så bruker vi den interne. Det er ikke en tung kalkulasjon.
 
 Satt litt fast på at det ikke fungerte, frem til jeg innså at storybook-eksemplene sender jo inn en verdi for selectedOptions. Den ble egentlig brukt som "intitalSelectedOptions", så det blir en liten funksjonell endring her.
 Jeg tror det er greit (sender man inn selectedOptions, så må man også håndtere oppdateringer av den), men vi kan evt. vurderere en "defaultSelectedOptions" a la "defaultValue" om noen spør etter det.

KenAJoh replied:
🙌 Hvis vi følger pattern som er brukt i de fleste andre komponenter gir `defaultSelectedOptions` mye mening. Hvis de setter `selectedOptions` selv må de også ta seg av oppdateringen som du nevner 👍

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -0,0 +1,212 @@
+
+.navds-combobox__wrapper-inner:has(.navds-combobox__input:focus).navds-combobox__wrapper-inner--virtually-unfocused {
+  box-shadow: none;
+}
```
Kan endres til `focus-visible` med fallback til `:focus` for ikke støttede browsere (se button.css)

it-vegard replied:
Done, men usikker på om det gir verdi. Ser ikke noen forskjell, og CSS-en blir dobbelt så stor.
Er ikke focus-visible mest for knapper, o.l.?

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -0,0 +1,212 @@
+  color: var(--a-text-action-selected);
+}
+
+.navds-combobox__button-toggle-list:focus {
```
Samme focus-visible kommentar som over

it-vegard replied:
Done, med samme kommentar som over.

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -0,0 +1,212 @@
+}
+
+.navds-combobox__list-item--focus,
+.navds-combobox__list-item:hover {
```
Bør list-item ha `cursor: pointer` på hover?

it-vegard replied:
Ikke natively, i hvert fall. Om vi ønsker, så er det ikke det verste å avvike på, men jeg pleier å bare følge native på slikt. (Sammenligner da med select + options)

Kan høre med Sjur i morgen.

it-vegard replied:
Har lagt til `cursor: pointer` på options etter prat med Sjur

KenAJoh commented on @navikt/core/css/form/combobox.css:
```diff
@@ -0,0 +1,212 @@
+  flex-direction: column;
+  align-items: flex-start;
+  padding: 0;
+  box-shadow: var(--a-shadow-large);
```
Shadow-large virker litt hefting nå, kan hende denne må justeres etter denne PR-en er merget https://github.com/navikt/aksel/pull/2041

it-vegard replied:
Endret til shadow-small. Det så ut til å stemme mer med designet.

it-vegard committed:
Differentiate between aria-autocomplete values "both" and "list" for combobox with and without autocomplete, respectively.

KenAJoh commented on @navikt/core/react/src/form/combobox/ComboboxProvider.tsx:
```diff
@@ -0,0 +1,63 @@
+import { ComboboxProps } from "./types";
+
+const ComboboxProvider = forwardRef<HTMLInputElement, ComboboxProps>(
+  (props, ref) => {
```
Ref blir ikke sendt videre til input

it-vegard replied:
Fixed

it-vegard committed:
Re-arrange props for ComboboxProps according to 1) required/optional and 2) alphabetical order

KenAJoh committed:
:test_tube: La inn interaksjonstest

KenAJoh committed:
:test_tube: La til state-tester i jest

it-vegard committed:
Add JSDoc to ComboboxProps type

Deleted the children prop, as we don't use it.

it-vegard committed:
Add an example of a controlled Combobox (using value prop)

Currently does not work properly, as filteredOptions does not update.

it-vegard committed:
Add support for defaultValue prop to Combobox

it-vegard committed:
Add combobox.css to _mappings.js

it-vegard committed:
Remove "undefined"/"undefinedagain" strings from list of options in storybook. They don't fit in.

it-vegard committed:
Remove unneccessary comment

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Move focusInput and code for selecting autocompleted text input inputContext, so we can reduce the number of useEffects

it-vegard committed:
Replace useEffect for checking if externalSelectedOptions has changed with just selecting which to use during render.

This causes a functional change from using selectedOptions as an "initital selected options" prop,
to forcing a controlled component where the app has to update the value as well, but that seems like a good API design.

it-vegard committed:
Activate input field when clicking anywhere inside of the border (including space around chips, etc.

it-vegard committed:
Merge branch 'main' into combobox

# Conflicts:
#	@navikt/core/css/expansioncard.css

JulianNymark reviewed (commented):
Er det rart at jeg ikke kan trykke 'enter / return' her (med ferdig utfylt verdi) og den blir ikke lagt til?

![image](https://github.com/navikt/aksel/assets/819074/ae9c724e-0bc3-4774-9f6e-c9afccb87ebf)

JulianNymark commented on @navikt/core/react/src/form/combobox/combobox.stories.tsx:
```diff
@@ -0,0 +1,319 @@
+      </div>
+    );
+  },
+  play: async ({ canvasElement }) => {
```
![image](https://github.com/navikt/aksel/assets/819074/62bad0fc-05d8-4a85-ac6d-df579f53d22c)

Denne storybook interactions/play skal vel ikke feile? (skriver 'apple' og finner ikke banana?)

KenAJoh replied:
Skal feile siden en keydown "esc" bør resette listen slik at neste gang man trykker "arrowDown" så vil alle valgene vises. banana er da først i listen med alle valg. Men kan hende et dette er tiltenkt interaksjon 🤔

JulianNymark replied:
hmm, må ha vært en kode endring kanskje? ser ikke ut til å virke her for meg (FF on macos), den blir tom (esc clearer input, men listen er fortsatt filtrert... very mystisk!)

it-vegard replied:
Nice catch! Ser ut til å være knyttet til innføringen av "searchTerm"-konseptet. Dette er en slags speiling av "value" som vi innførte med single select-koden. Den oppdateres bare når brukeren skriver i input-feltet, og ikke når teksten autocompletes. Dermed har vi brukt både setValue og setSearchTerm flere steder. Ser ut som vi glemte det i handleClear-funksjonen.

Hadde sett denne feilen tidligere, men glemte den. Bra å få inn tester på det!

Nå kjører storybook-testene, men ikke RTL-testene. Prøver å debugge hvorfor. Håper det er noe caching-greier, for testene er jo omtrent like... 😆

it-vegard replied:
Ser ut som koden for escape er `{Escape}`, ikke `{esc}`. Da kjører RTL-testen også. :)

JulianNymark reviewed (commented):
om jeg tar pil ned og opp 🤔 ser ut som jeg kan trigge strange behaviour.

https://github.com/navikt/aksel/assets/819074/ea36446c-dde7-4d0a-85ea-65a8ed005726

JulianNymark reviewed (commented):
https://github.com/navikt/aksel/assets/819074/dea2b273-ea09-4b8b-916f-9eae2075c65d

Ser ut som at jeg kan velge en "hidden add" option, og kan legge til custom elementer selv om `allowNewValues` er false

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
Merge remote-tracking branch 'origin/main' into combobox

it-vegard committed:
Update searchTerm when clearing the input value. Should have been done when adding single select, but missed this one.

it-vegard committed:
Use correct naming for escape key code. I think "esc" is the "keyCode", not "code". "keyCode" is deprecated.

it-vegard committed:
Should not be able to toggle an option while the dropdown is closed

it-vegard committed:
Revert printWidth from 120 and back to 130, to avoid diffs for several unrelated files.

If we want to make this change, we should do it in another PR.

it-vegard committed:
Set single select as default for Combobox. Should probably add more specific test for either single- or multi-select

it-vegard committed:
Update Storybook-examples

- Missing shouldAutoComplete on single-select example (should this be default?) and isMultiSelect on external chips example
- Deleted redundant single select examples
- Created a proper multi-select example
- Default open list for example with no search hits

it-vegard committed:
Re-arrange combobox storybook examples to better match expected use

it-vegard committed:
Delete unnecessary example

it-vegard committed:
Add filteredOptions as a prop, to allow for async/external search.

it-vegard committed:
Merge branch 'main' into combobox

it-vegard committed:
Use :focus-visible with fallback to :focus

it-vegard committed:
Lesser shadow

it-vegard committed:
Added simple JSDoc for Combobox

it-vegard committed:
Remove unused exported function and redundant useEffect used to update state when props change

The useEffect is redundant, because the same update happens in the isListOpen useMemo further up in the code

it-vegard committed:
Remove clear button in Combobox from tab order. It should still be clickable, but this functionality is handled by pressing the escape key instead.

Fewer tab stops make for a better user experience.

it-vegard committed:
Handle possible edge case where currentOption === ""

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Remove comments

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

it-vegard committed:
Pass external ref down to the Combobox component, where it will be merged with our internal inputRef

it-vegard committed:
Fix bug where you could end up adding options even if allowNewValues is false

it-vegard committed:
Safeguards against [].length - 1 === -1, causing us to add new values even if not allowed

JulianNymark committed:
add some storybook tests to combobox

it-vegard committed:
Open filteredOptions-dropdown on focusing the combobox input field

it-vegard committed:
Merge branch 'main' into combobox

# Conflicts:
#	yarn.lock

it-vegard committed:
Add readOnly to FormFieldType

it-vegard committed:
Move logic for handling toggleOption into selectedOptionsContext, and letting it handle customOptions as well

The goal is to make the Combobox component much simpler, and making the logic of combining customOptions and selectedOptions more transparent

it-vegard committed:
Select value when pressing 'enter' in Combobox with autocomplete, and show selected value as an unstyled Chip instead of just text

Makes the text be removable by pressing backspace, and eliminates issue with changing a "selected" value.

it-vegard committed:
Fix issue with not being able to click on options + list re-opening after clicking the "toggle list open"-button

Turns out the onBlur on input closed the dropdown before being able to register the click on option, and
onFocus on input re-opened the dropdown after clicking the toggle-button.

it-vegard commented on @navikt/core/react/src/form/combobox/types.ts:
```diff
@@ -0,0 +1,135 @@
+   * @param event
+   * @returns
+   */
+  onClear?: (event: ComboboxClearEvent) => void;
```
Burde diskutere typer for onXXX-funksjonene:

- Skal alle være native events som sendes videre?
- Bør vi bare sende med verdier som kan være nyttig å ha her, som hvilken verdi som fjernes/endres/etc?

it-vegard committed:
Properly clear input field (not for multiselect) and re-focus the input in all cases where we toggle an option

Removes edge-cases where we lost focus or ended up with the value not cleared when selecting an option with a pointer

it-vegard committed:
Always pass the event object to clearInput, so that the callback function can receive it

it-vegard committed:
Add cursor: pointer to combobox options list

Wanted to do same change for select options, but those doesn't seem like native options can have even the cursor be styled

it-vegard committed:
Update type for Combobox onClear function to reflect that it is now passed a real event

it-vegard committed:
Replace fireEvent.click with userEvent.click because it doesn't trigger pointer events

it-vegard committed:
Add act wrappers to account for state change (open list) when clicking the combobox input field

it-vegard committed:
Prevent opening list onFocus when focus is set to ToggleListButton, as this will cause a double-click event, which will open and then close the list immediately

it-vegard committed:
Fix incorrect ref type

it-vegard committed:
If a value is selected in the dropdown/FilteredOptions, that should override check against trying to add an existing value.

Otherwise this would prevent the user from searching for the value that has been selected, and then unselect it.

it-vegard committed:
Unnecessary to add size="medium", as it is the default for BodyShort

it-vegard committed:
Typing out the full name of the variable, because it is easier to read

it-vegard committed:
Since the "add new option"-button will never actually be selected, just set aria-selected to false

it-vegard committed:
Controlled example has multiple selected values, and should therefore be a multiselect

it-vegard committed:
Reduce flimmering of filtered list during typing, where it would update the list just before closing it

Move "calculation" of ariaDescribedby to a useMemo instead of useEffect, and use useLayoutEffect instead of useEffect to update isListOpen synchronously before re-rendering.
Will look into moving the toggleIsListOpen to an event listener instead.

it-vegard committed:
Toggle list open/closed based on input value in the onChange listener instead of a useLayoutEffect

larseirikhansen committed:
combobox css tokens

larseirikhansen committed:
fix colors combobox css tokens

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

it-vegard committed:
Remove hover effect from loading spinner and only show cursor:pointer on actual, clickable options

it-vegard committed:
Better text for loading state

it-vegard committed:
Remove background-color for selected-options: single select should have no background, and multiselect uses chips with their styling

it-vegard committed:
Merge branch 'main' into combobox

it-vegard committed:
Move "add new" button above "no search hits" text to avoid "jumping" when going from search hits to no search hits

it-vegard committed:
Adjustments according to Figma sketches
- Background color
- Padding
- Icon size

larseirikhansen committed:
css token left borders

it-vegard committed:
Add examples with both single- and multiselect, both in medium and small sizes.

it-vegard committed:
Properly place combobox dropdown over underlying objects and make sure the entire background is colored white.

it-vegard committed:
Remove hover effect on single select. Styling for multiselect is handled by Chips component.

it-vegard committed:
Close dropdown when selecting a value for single select, to hopefully avoid "jumping" keyboard and component

it-vegard committed:
Merge branch 'main' into combobox

# Conflicts:
#	yarn.lock

it-vegard committed:
Prevent autocompleting when only entering space

it-vegard committed:
Just trying out useLayoutEffect as a temporary fix for Android issue where autocomplete breaks and adds the beginning of the word after a few letters have been typed

it-vegard committed:
Try to fix select-on-enter on mobile keyboard

it-vegard committed:
Another attempt at reacting to the "Done" key (should have triggered enter) on mobile keyboard

it-vegard committed:
Set enter as key hint for "done"/"OK"/"enter" key on mobile

it-vegard committed:
Trying out search input mode to see if that stops the bug where the field is cleared on enter/OK on mobile

it-vegard committed:
Finally fixed issue where we thought we couldn't select autocompleted text using "enter" on mobile keyboards. The actual issue was that capitalized words broke an equality check, causing us to try to add a new value. Since that wasn't active on single select example, we just cleared the field and moved on.

it-vegard committed:
Make combobox input field flow more naturally, expanding the available width and just breaking to a new line when the space becomes too small.

it-vegard committed:
Make combobox chips smaller for size="small"

it-vegard committed:
Proper padding in the "input field" (bordered region) for size="medium" and size="small"

it-vegard committed:
Proper sizing for FilteredOptions

it-vegard committed:
Proper sizing and placement for clear- and toggleList-buttons

Should be last adjustment for size="small"

it-vegard committed:
Don't hard-code value in example. That was just used during development.

larseirikhansen committed:
Merge remote-tracking branch 'origin/combobox' into combobox

larseirikhansen committed:
css tokens bg fix, add item borders

larseirikhansen committed:
remove unnecessary comment

larseirikhansen committed:
plusicon aria-hidden

larseirikhansen committed:
fix input overlapping clear and toggleButton

larseirikhansen committed:
fixed: clicking in gap between li-elements (ul) closed list

larseirikhansen committed:
add UNSAFE_ prefix to combobox

larseirikhansen committed:
changeset minor update.md

larseirikhansen committed:
code examples for docs
